nRF9160 pca10090

ioannisg opened the pull request:
Introducing nRF9160_PCA10090 board

PR over #12531 

Umbrella issue: #12199

---

Update: since #12531 is merged, this PR contains only the nRF9160 DK board definition.

ioannisg committed:
boards: arm: nrf: board definition for nrf9160_pca10090

This commit contributes the Kconfig and DTS files for the
nrf9160_pca10090 board definition. The nrf9160_pca10090 board
shall be used to build Zephyr applications, running in Secure
mode, on nRF9160_PCA10090 Dev Kit.

Additionally, the commit introduces a default Secure and
Non-Secure flash partition configuration for the
nRF9160_PCA10090 board, allowing to use Zephyr for building
both the Secure and Non-Secure firmware images.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: nrf: board definition for nrf9160_pca10090 (non-secure)

This commit contributes the Kconfig and DTS board definition
files for nrf9160_pca10090ns board. This Zephyr board shall be
used to build Non-Secure Zephyr applications on nrf9160_pca10090
Dev Kit.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

SebastianBoe commented on boards/arm/nrf9160_pca10090/board.cmake:
```diff
@@ -0,0 +1,4 @@
+board_runner_args(nrfjprog "--nrf-family=NRF91" "--softreset")
```
Should be a comment explaining why we should have softreset on nRF91, but not on most other nrf platforms.

```
nrf52810_pca10040/board.cmake
1:board_runner_args(nrfjprog "--nrf-family=NRF52")

nrf52_pca10040/board.cmake
1:board_runner_args(nrfjprog "--nrf-family=NRF52")

nrf52840_pca10059/board.cmake
1:board_runner_args(nrfjprog "--nrf-family=NRF52")

nrf52840_pca10056/board.cmake
1:board_runner_args(nrfjprog "--nrf-family=NRF52")

nrf52_pca20020/board.cmake
1:board_runner_args(nrfjprog "--nrf-family=NRF52" "--softreset")

particle_xenon/board.cmake
8:board_runner_args(nrfjprog "--nrf-family=NRF52")

nrf51_ble400/board.cmake
1:board_runner_args(nrfjprog "--nrf-family=NRF51")

nrf9160_pca10090/board.cmake
1:board_runner_args(nrfjprog "--nrf-family=NRF91" "--softreset")

nrf52_adafruit_feather/board.cmake
1:board_runner_args(nrfjprog "--nrf-family=NRF52")

nrf51_pca10028/board.cmake
1:board_runner_args(nrfjprog "--nrf-family=NRF51")

particle_argon/board.cmake
7:board_runner_args(nrfjprog "--nrf-family=NRF52")

particle_boron/board.cmake
7:board_runner_args(nrfjprog "--nrf-family=NRF52")

```

ioannisg replied:
@SebastianBoe you are right on this. Actually, pin-reset seems to work on the latests kits, so we can remove this option. Will do.

ioannisg commented:
@lemrey @SebastianBoe @carlescufi @anangl ; requesting early feedback on _board_ definition

ioannisg commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090ns_defconfig:
```diff
@@ -0,0 +1,26 @@
+CONFIG_ARM_TRUSTZONE_M=y
+
+# This Board implies building Non-Secure firmware
+CONFIG_TRUSTED_EXECUTION_NONSECURE=y
```
This one is in accordance with v2m_musca, where the "non-secure" version of the board determines that we are building a NONSECURE firmware.

SebastianBoe commented:
> @lemrey @SebastianBoe @carlescufi @anangl ; requesting early feedback on board definition

Could you be more specififc, which files are in-scope, and which files are out-of-scope for feedback.

ioannisg commented:
> 
> 
> > @lemrey @SebastianBoe @carlescufi @anangl ; requesting early feedback on board definition
> 
> Could you be more specififc, which files are in-scope, and which files are out-of-scope for feedback.

Sure. I 'd like fb on /boards/arm/nrf9160_pca10090/*

SebastianBoe commented on boards/arm/nrf9160_pca10090/board.h:
```diff
@@ -0,0 +1,13 @@
+#include <soc.h>
+
+
+#endif /* __INC_BOARD_H */
```
It is uncommon to have a board.h file.

```
sebo@mach:~/ncs/zephyr/boards/arm$ find . -name board.h
find . -name board.h
./efr32_slwstk6061a/board.h
./efm32wg_stk3800/board.h
./reel_board/board.h
./nrf52_adafruit_feather/board.h
./bbc_microbit/board.h
./efm32hg_slstk3400a/board.h
./cc2650_sensortag/board.h
./efm32pg_stk3402a/board.h
./mps2_an385/board.h

```

and unprecedented to include soc.h from it.

So we need to either remove this file or come up with an explanation for what it is doing here.

lemrey replied:
I agree, this can be removed and we should include `soc.h` directly instead.

ioannisg replied:
Yes - will do this, thanks.

ioannisg replied:
Removed.

lemrey commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090_partition_conf.dts:
```diff
@@ -0,0 +1,62 @@
+ */
+
+&flash0 {
+	reg = <0x00000000 DT_SIZE_K(256)>;
```
This is probably wrong, we are partition a space of one megabyte. Could you please double check?

ioannisg replied:
Yes, this is deliberate; this is a Secure nrf9160_pca10090 board - and requires to save some space for the non-secure image. If no-MCU boot is present, then the first 256 KB are allocated to the Secure image.

Hopefully we are going to address the problem of partitioning, properly, together with multi-image building

lemrey replied:
This will limit the amount of flash available to secure firmware to 256k, but assumes that a non-secure firmware image is also present, which might note be the case.

hakonfam replied:
I agree that this is suitable for now. Future solutions will make this more configurable.

ioannisg replied:
Then let's leave this for now and address it properly in the context of multi-image builds.

lemrey commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090_partition_conf.dts:
```diff
@@ -0,0 +1,62 @@
+ */
+
+/*
+ * Default Flash planning for nRF9160_pca10090.
```
What about the storage partition, are we going to add that a later point or?

lemrey replied:
Nevermind, I see now that it is defined in another file.

lemrey commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090.yaml:
```diff
@@ -0,0 +1,7 @@
+identifier: nrf9160_pca10090
+name: nRF9160-PCA10090
+type: mcu
+arch: arm
```
Should we add the RAM and FLASH properties?
`ram: 256 flash: 1024`

ioannisg replied:
We do not do this, consistently, for nRF boards. But I can do that for nrf9160_pca10090

ioannisg replied:
This is now done.

lemrey commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090ns.yaml:
```diff
@@ -0,0 +1,7 @@
+identifier: nrf9160_pca10090ns
```
Should we add the RAM and FLASH properties?
`ram: 256 flash: 1024`

ioannisg replied:
Same as above. I will add them.

ioannisg replied:
@lemrey @SebastianBoe , what values, exactly, do we need to add here?
- I added RAM=64K and FLASH=256K, because _this_ is what the current board definition allows by default. These are different from the SOCs actual resources, though

SebastianBoe replied:
It is used to filter tests that require a lot of ram.

So it must match what the board definition allows.

ioannisg replied:
Then this is fine.

SebastianBoe commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090_common.dts:
```diff
@@ -0,0 +1,151 @@
+};
+
+/* Include partition configuration file */
+#include "nrf9160_pca10090_partition_conf.dts"
```
I'm not seeing the need / desire for splitting up these two files.

ioannisg replied:
For nRF9160 Board, the chosen default partitioning is complex and comes with lots of explanatory comment, so I chose to put it in a separate file, to remove the extra "noise" from nrf9160_pca10090_common.dts.

If you insist, I can merge the two files; no strong opinion on this.

SebastianBoe reviewed (approved):
Generally LGTM.

ioannisg commented on boards/arm/nrf9160_pca10090/board.cmake:
```diff
@@ -0,0 +1,4 @@
+board_runner_args(nrfjprog "--nrf-family=NRF91")
```
pin-reset is working for nrf9160_pca10090, so I modified this, @SebastianBoe , to align with the rest of the Nordic Dev-Kits

ioannisg replied:
Resolving this one (was only to make @SebastianBoe aware that the runner-args are now the default nrf board runner args)

lemrey commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090ns.dts:
```diff
@@ -0,0 +1,18 @@
+	chosen {
+		zephyr,flash = &flash0;
+		zephyr,sram = &sram0_ns;
+		zephyr,code-partition = &slot0_ns_partition;
```
How does this play with `dts/common/mcuboot.overlay` that specifies
`zephyr,code-partition = &slot0_partition;`
What is the precedence?

SebastianBoe replied:
That is correct.

lemrey replied:
@ioannisg this thread has not been resolved 🐛

ioannisg replied:
MCUboot should not be allowed for Non-Secure images - this is something we could agree on, right? [Sorry, mistake to resolve that one]

lemrey replied:
We cleared this up. For now this is fine as it is, but non-secure firmware images that wish to be programmed using MCUboot need to leave the "MCUBoot support" option unchecked, otherwise MCUBoot overlay will override the code partition and they will be linked in the secure one.

ioannisg commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090.yaml:
```diff
@@ -0,0 +1,9 @@
+toolchain:
+  - gnuarmemb
+  - xtools
+ram: 64
```
Same comment as for the Non-Secure board; I added what is actually available by default with this board definition.

dbkinder reviewed (commented):
No docs to review here (yet)...

anangl commented on boards/arm/nrf9160_pca10090/Kconfig.defconfig:
```diff
@@ -0,0 +1,12 @@
+
+config BOARD
+	default "nrf9160_pca10090"
+
```
Please add here:
```
if ADC

config ADC_0
	default y

endif # ADC

if I2C

config I2C_0
	default y

endif # I2C

if SPI

config SPI_1
	default y

endif # SPI
```
to enable default instances, like it is done for other Nordic DK boards.
This will allow users to build samples and tests that use these peripherals without extra configuration.

ioannisg replied:
Thanks. Will do

ioannisg replied:
Done

anangl commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090_defconfig:
```diff
@@ -0,0 +1,23 @@
+
+# Enable UARTE
+CONFIG_SERIAL=y
+CONFIG_UART_NRFX=y
```
This is redundant (the option is enabled by default).

ioannisg replied:
Thanks, done. And for the "ns" fail, too.

anangl commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090.dts:
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
```
I guess this line should be removed.

ioannisg replied:
Yes, thanks. Done.

anangl commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090_common.dts:
```diff
@@ -0,0 +1,151 @@
+	rts-pin = <14>;
+	cts-pin = <15>;
+};
+
```
Please, add:
```
&adc {
	status ="ok";
};

&i2c0 {
	status = "ok";
	sda-pin = <30>;
	scl-pin = <31>;
};

&spi1 {
	status = "ok";
	sck-pin = <19>;
	mosi-pin = <18>;
	miso-pin = <17>;
};
```

anangl replied:
> What would happen if a project overlay tried to assign those pins to something else? We have a project where we use them for UART, for example.

You can always reconfigure the pins for this SPI instance (or disable it completely) in the overlay when you need them for a different purpose. This configuration will serve as a starting point only. I wanted to make it similar to what is done for other DKs. But we can use different pins for this SPI. Could you suggest some?

ioannisg replied:
I would suppose, @anangl @lemrey we have the same problem with every project that overwrites default PIN configuration, no?

lemrey replied:
@anangl I understand, that shouldn't be a problem then if it can be overridden without errors.

ioannisg replied:
Then I will add the proposed pin configuration in the board definition.

ioannisg replied:
Done

ioannisg commented:
> 
> 
> No docs to review here (yet)...

@dbkinder I will notify you upon uploading an .rst for nrf9160_pca10090

jtguggedal commented on dts/arm/nordic/nrf9160.dtsi:
```diff
@@ -28,10 +28,11 @@
 	aliases {
 		flash-controller = &flash_controller;
 		ficr = &ficr;
+		rtc-0 = &rtc0;
```
Could we add PWM peripherals as well? 
For reference: https://github.com/barsok/fw-nrfconnect-zephyr/commit/f9385b4b19fca1f0455af079e9e700454390af02

anangl replied:
Good point. I'll add them in #12531.

ioannisg replied:
They are added. Resolving this one.

ioannisg committed:
boards: nrf: add initial documentation for nrf9160_pca10090

This commit adds the initial documentation information
for board nrf9160_pca10090.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented on boards/arm/nrf9160_pca10090/doc/nrf9160_pca10090.rst:
```diff
@@ -0,0 +1,106 @@
+     nRF9160 PCA10090 DK (Credit: Nordic Semi)
+
+More information about the board can be found at the
+`nRF91 DK website`_. The `Nordic Semiconductor Documentation library`_
```
Still working on this, thanks.

ioannisg commented on boards/arm/nrf9160_pca10090/doc/nrf9160_pca10090.rst:
```diff
@@ -0,0 +1,106 @@
+------------
+
+* BUTTON1 = SW1 = P0.6
+* BUTTON2 = SW2 = P0.9
```
@anangl : Push Buttons (1,2) seem to be connected to P0.08 and P0.09, AFAIK.
Switches 1,2 are connected to 6,7

ioannisg replied:
Resolving this one, since the pin numbers are fixed.

ioannisg commented:
@dbkinder, @carlescufi  could you review the /doc material (last commit in this PR?

FYI  @agross-linaro , @galak Could you take a look? I have tried to align the nRF9160 Board Support with the v2m Musca, i.e. adding both a default(Secure) and a Non-Secure nrf9160 boards in the same boards' sub-directory.

agross-oss reviewed (approved):
looks fine

ioannisg commented:
FYI @jhn-nordic

ioannisg committed:
CODEOWNERS: Code-owner for nRF9160 Board

Assigning code-owner for Nordic Semiconductor nRF9160_pca10090
development board, featuring nRF9160 and ARMv8-M core with
ARM TrustZone-M.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

carlescufi commented on boards/arm/nrf9160_pca10090/doc/nrf9160_pca10090.rst:
```diff
@@ -0,0 +1,206 @@
+* :abbr:`SPI (Serial Peripheral Interface)`
+* :abbr:`UART (Universal asynchronous receiver-transmitter)`
+* :abbr:`WDT (Watchdog Timer)`
+
```
We should add `TEE` here

ioannisg replied:
Right; i added SPU in the table below. Perhaps i could add IDAU here (generic term)??

ioannisg replied:
Added IDAU in this table.

carlescufi commented on boards/arm/nrf9160_pca10090/doc/nrf9160_pca10090.rst:
```diff
@@ -0,0 +1,206 @@
++-----------+------------+----------------------+
+| WDT       | on-chip    | watchdog             |
++-----------+------------+----------------------+
+
```
We should add `TEE` here

ioannisg replied:
SPU is added

carlescufi commented on boards/arm/nrf9160_pca10090/doc/nrf9160_pca10090.rst:
```diff
@@ -0,0 +1,206 @@
+
+The process requires the following steps:
+
+1. Build the Secure Zephyr application using ``-DBOARD=nrf9160_pca10090`` and
```
We probably need a bit more doc about TZ and how you can also combine Zephyr with TF-M. This should probably be elsewhere, somewhere in `doc/`. No need for this to be done here, it can be in a follow-up PR.

ioannisg replied:
I agree on this. Perhaps we can do this together with @agross-linaro

carlescufi reviewed (approved):
Looks good enough to me at this point. Additional doc is required in `doc/`, but this is not part of this PR, as the Musca board is already merged and it already supports Armv8-M.

nvlsianpu commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090.yaml:
```diff
@@ -0,0 +1,9 @@
+  - gnuarmemb
+  - xtools
+ram: 64
+flash: 256
```
EOL

ioannisg replied:
Fixed, thanks.

ioannisg commented:
@dbkinder Could you review the documentation for nrf9160_pca10090 (`./doc/nrf9160_pca10090.rst`)?

ioannisg commented on boards/arm/nrf9160_pca10090/doc/nrf9160_pca10090.rst:
```diff
@@ -0,0 +1,206 @@
+* :abbr:`RTC (nRF RTC System Clock)`
+* Segger RTT (RTT Console)
+* :abbr:`SPI (Serial Peripheral Interface)`
+* :abbr:`UART (Universal asynchronous receiver-transmitter)`
```
Right - I am fixing this, thanks.

galak reviewed (changes requested):
Seeing some sanitycheck issues, please fix:

`./scripts/sanitycheck -p nrf9160_pca10090`

```
nrf9160_pca10090          tests/benchmarks/timing_info/benchmark.timing.userspace FAILED: build_error
	see: sanity-out/nrf9160_pca10090/tests/benchmarks/timing_info/benchmark.timing.userspace/build.log

total complete:   28/ 158  17%  failed:    1

nrf9160_pca10090          samples/subsys/logging/logger/samples.logger       FAILED: build_error
	see: sanity-out/nrf9160_pca10090/samples/subsys/logging/logger/samples.logger/build.log

total complete:   29/ 158  18%  failed:    2

nrf9160_pca10090          tests/benchmarks/timing_info/benchmark.timing.default_kernel FAILED: build_error
	see: sanity-out/nrf9160_pca10090/tests/benchmarks/timing_info/benchmark.timing.default_kernel/build.log

total complete:   30/ 158  18%  failed:    3

nrf9160_pca10090          samples/subsys/shell/shell_module/test_minimal     FAILED: build_error
	see: sanity-out/nrf9160_pca10090/samples/subsys/shell/shell_module/test_minimal/build.log

total complete:   42/ 158  26%  failed:    4

nrf9160_pca10090          samples/mpu/mpu_test/test                          FAILED: build_error
	see: sanity-out/nrf9160_pca10090/samples/mpu/mpu_test/test/build.log

total complete:   43/ 158  27%  failed:    5

nrf9160_pca10090          samples/subsys/shell/shell_module/test             FAILED: build_error
	see: sanity-out/nrf9160_pca10090/samples/subsys/shell/shell_module/test/build.log

total complete:   56/ 158  35%  failed:    6

nrf9160_pca10090          tests/drivers/uart/uart_basic_api/peripheral.uart.shell FAILED: build_error
	see: sanity-out/nrf9160_pca10090/tests/drivers/uart/uart_basic_api/peripheral.uart.shell/build.log

```

galak commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090.yaml:
```diff
@@ -0,0 +1,9 @@
+  - gnuarmemb
+  - xtools
+ram: 64
+flash: 256
```
Seems like we should have some list of 'supported' devices here.

ioannisg replied:
I can try to add a list here (not consistent among ARM or even NRF boards, though)

galak reviewed (changes requested):
Can you fix the issues related to these dtc warnings, we should only have one of the 2 devices enabled by default.

```
nrf9160_pca10090.dts.pre.tmp:100.18-106.3: Warning (unique_unit_address_if_enabled): /soc/peripheral@50000000/uart@8000: duplicate unit-address (also used in node /soc/peripheral@50000000/i2c@8000)
  also defined at nrf9160_pca10090.dts.pre.tmp:391.8-398.3
nrf9160_pca10090.dts.pre.tmp:107.18-113.3: Warning (unique_unit_address_if_enabled): /soc/peripheral@50000000/uart@9000: duplicate unit-address (also used in node /soc/peripheral@50000000/spi@9000)
  also defined at nrf9160_pca10090.dts.pre.tmp:399.8-406.3
nrf9160_pca10090.dts.pre.tmp:252.19-257.3: Warning (unique_unit_address_if_enabled): /soc/peripheral@50000000/clock@5000: duplicate unit-address (also used in node /soc/peripheral@50000000/power@5000)
-
```

galak commented:
```
[galak@spiff zephyr]$ ./scripts/sanitycheck -p nrf9160_pca10090ns
JOBS: 64
Cleaning output directory /home/galak/git/zephyr/sanity-out
Building testcase defconfigs...
161 tests selected, 89194 tests discarded due to filters
total complete:   23/ 159  14%  failed:    0

nrf9160_pca10090ns        samples/subsys/logging/logger/samples.logger       FAILED: build_error
	see: sanity-out/nrf9160_pca10090ns/samples/subsys/logging/logger/samples.logger/build.log

total complete:   25/ 159  15%  failed:    1

nrf9160_pca10090ns        tests/benchmarks/timing_info/benchmark.timing.default_kernel FAILED: build_error
	see: sanity-out/nrf9160_pca10090ns/tests/benchmarks/timing_info/benchmark.timing.default_kernel/build.log

total complete:   26/ 159  16%  failed:    2

nrf9160_pca10090ns        samples/subsys/shell/shell_module/test_minimal     FAILED: build_error
	see: sanity-out/nrf9160_pca10090ns/samples/subsys/shell/shell_module/test_minimal/build.log

total complete:   29/ 159  18%  failed:    3

nrf9160_pca10090ns        tests/benchmarks/timing_info/benchmark.timing.userspace FAILED: build_error
	see: sanity-out/nrf9160_pca10090ns/tests/benchmarks/timing_info/benchmark.timing.userspace/build.log

total complete:   32/ 159  20%  failed:    4

nrf9160_pca10090ns        samples/mpu/mpu_test/test                          FAILED: build_error
	see: sanity-out/nrf9160_pca10090ns/samples/mpu/mpu_test/test/build.log

total complete:   36/ 159  22%  failed:    5

nrf9160_pca10090ns        samples/subsys/shell/shell_module/test             FAILED: build_error
	see: sanity-out/nrf9160_pca10090ns/samples/subsys/shell/shell_module/test/build.log

total complete:   61/ 159  38%  failed:    6

nrf9160_pca10090ns        tests/drivers/uart/uart_basic_api/peripheral.uart.shell FAILED: build_error
	see: sanity-out/nrf9160_pca10090ns/tests/drivers/uart/uart_basic_api/peripheral.uart.shell/build.log

total complete:  159/ 159  100%  failed:    7
152 of 159 tests passed with 0 warnings in 144 seconds
```

ioannisg commented:
> 
> 
> Can you fix the issues related to these dtc warnings, we should only have one of the 2 devices enabled by default.
> 
> ```
> nrf9160_pca10090.dts.pre.tmp:100.18-106.3: Warning (unique_unit_address_if_enabled): /soc/peripheral@50000000/uart@8000: duplicate unit-address (also used in node /soc/peripheral@50000000/i2c@8000)
>   also defined at nrf9160_pca10090.dts.pre.tmp:391.8-398.3
> nrf9160_pca10090.dts.pre.tmp:107.18-113.3: Warning (unique_unit_address_if_enabled): /soc/peripheral@50000000/uart@9000: duplicate unit-address (also used in node /soc/peripheral@50000000/spi@9000)
>   also defined at nrf9160_pca10090.dts.pre.tmp:399.8-406.3
> nrf9160_pca10090.dts.pre.tmp:252.19-257.3: Warning (unique_unit_address_if_enabled): /soc/peripheral@50000000/clock@5000: duplicate unit-address (also used in node /soc/peripheral@50000000/power@5000)
> -
> ```

Not sure I understand this @galak .
For example: CLOCK and POWER need to be both enabled.

I though these warning would go of after #12590. Am I missing something?

galak commented:
> Not sure I understand this @galak .
> For example: CLOCK and POWER need to be both enabled.
> 
> I though these warning would go of after #12590. Am I missing something?

Let me look at the CLOCK/POWER case.  The uart/i2c, and uart/spi should be addressed by only having one enabled by default in the dts.

ioannisg commented:
> 
> 
> > Not sure I understand this @galak .
> > For example: CLOCK and POWER need to be both enabled.
> > I though these warning would go of after #12590. Am I missing something?
> 
> Let me look at the CLOCK/POWER case. The uart/i2c, and uart/spi should be addressed by only having one enabled by default in the dts.

Thanks @galak 
Now - in nRF9160, the usual macros defining the peripheral base addresses (e.g. NRF_CLOCK, NRF_TIMER_0, etc.) are **_not_** defined in the Nordic MDK headers. Their base addresses are _only_ defined in DTS format in SOC DTS (I assume this is good news), and we have #defines in nrfx_config_nrf9160.h that do the final mapping, e.g.

```
#ifdef DT_NORDIC_NRF_TIMER_TIMER_0_BASE_ADDRESS
#define NRF_TIMER0 \
        ((NRF_TIMER_Type *)DT_NORDIC_NRF_TIMER_TIMER_0_BASE_ADDRESS)
#endif

``` 

For the DTS- defines to be present, the device nodes in DTS shall have "status OK".

If we want the basic peripheral examples in the tree with I2C, serial, SPI, etc. to compile with the default nrf9160 board, we need to have all these enabled ("status OK"), as far as I understand. 

Therefore, we need to accept that the board definition will "enable" devices that (in practice) cannot be used simultaneously. Do you think there is another way  to work around the problem?

ioannisg committed:
soc: nrf: add include of nrfx_config_nrf9160.h in nrf9160 soc.h

In soc.h of nRF9160 SoC we need to explicitly include
nrfx_config_nrf9160.h to have the standard peripheral
base address macro mappings in Zephyr builds for nRF9160
SoC. (We need to do this spefically for nRF9160, as those
macros are not present in the nRF9160 MDK header).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
fYI @pdunaj

galak commented:
> If we want the basic peripheral examples in the tree with I2C, serial, SPI, etc. to compile with the default nrf9160 board, we need to have all these enabled ("status OK"), as far as I understand.
> 
> Therefore, we need to accept that the board definition will "enable" devices that (in practice) cannot be used simultaneously. Do you think there is another way to work around the problem?

However, if the define's are generated they should be used either.  So it shouldn't be an issue if `NRF_TIMER0` is defined as:
```#define NRF_TIMER0 \
        ((NRF_TIMER_Type *)DT_NORDIC_NRF_TIMER_TIMER_0_BASE_ADDRESS)
#endif
```

Since if TIMER0 isn't in the DTS than NRF_TIMER0 shouldn't be used in any code we're compiling.  Or am I missing something?

ioannisg commented:
>However, if the define's are generated they should be used either. So it shouldn't be an issue if NRF_TIMER0 is defined as:

        ((NRF_TIMER_Type *)DT_NORDIC_NRF_TIMER_TIMER_0_BASE_ADDRESS)

>Since if TIMER0 isn't in the DTS than NRF_TIMER0 shouldn't be used in any code we're compiling. Or am I missing something?

@galak this is exactly what is happening :) 
But for `DT_NORDIC_NRF_TIMER_TIMER_0_BASE_ADDRESS `to exist, we need to have "Status OK" on the DTS node for timer_0.

...and the same for spi0, i2c0, uarte0 DTS nodes, if we want the respective examples to compile.

But _that_ triggers the duplicate address warnings you mentioned above.

ioannisg commented:
> 
> 
> Seeing some sanitycheck issues, please fix:
> 
> `./scripts/sanitycheck -p nrf9160_pca10090`
> 
> ```
> nrf9160_pca10090          tests/benchmarks/timing_info/benchmark.timing.userspace FAILED: build_error
> 	see: sanity-out/nrf9160_pca10090/tests/benchmarks/timing_info/benchmark.timing.userspace/build.log
> 
> total complete:   28/ 158  17%  failed:    1
> 
> nrf9160_pca10090          samples/subsys/logging/logger/samples.logger       FAILED: build_error
> 	see: sanity-out/nrf9160_pca10090/samples/subsys/logging/logger/samples.logger/build.log
> 
> total complete:   29/ 158  18%  failed:    2
> 
> nrf9160_pca10090          tests/benchmarks/timing_info/benchmark.timing.default_kernel FAILED: build_error
> 	see: sanity-out/nrf9160_pca10090/tests/benchmarks/timing_info/benchmark.timing.default_kernel/build.log
> 
> total complete:   30/ 158  18%  failed:    3
> 
> nrf9160_pca10090          samples/subsys/shell/shell_module/test_minimal     FAILED: build_error
> 	see: sanity-out/nrf9160_pca10090/samples/subsys/shell/shell_module/test_minimal/build.log
> 
> total complete:   42/ 158  26%  failed:    4
> 
> nrf9160_pca10090          samples/mpu/mpu_test/test                          FAILED: build_error
> 	see: sanity-out/nrf9160_pca10090/samples/mpu/mpu_test/test/build.log
> 
> total complete:   43/ 158  27%  failed:    5
> 
> nrf9160_pca10090          samples/subsys/shell/shell_module/test             FAILED: build_error
> 	see: sanity-out/nrf9160_pca10090/samples/subsys/shell/shell_module/test/build.log
> 
> total complete:   56/ 158  35%  failed:    6
> 
> nrf9160_pca10090          tests/drivers/uart/uart_basic_api/peripheral.uart.shell FAILED: build_error
> 	see: sanity-out/nrf9160_pca10090/tests/drivers/uart/uart_basic_api/peripheral.uart.shell/build.log
> ```

Thanks @galak for this report. I applied the following fixes:
- added timer_2 in the list of enabled nodes in the pca10090 board (so I can get the NRF_TIMER2 macro defined, as discussed above). This fixes the timing_info tests
- added the shell-uart in the "chosen" list in board .dts
- modified soc.h (nrf9160) so it explicitly includes nrfx_config_nrf9160.h (separate commit on top)

ru-fu commented on boards/arm/nrf9160_pca10090/doc/nrf9160_pca10090.rst:
```diff
@@ -0,0 +1,207 @@
+********
+
+Zephyr applications use the nrf9160_pca10090 board configuration
+to run on the nRF9160 Development Kit (PCA10090) hardware. The
```
The official name is "nRF9160 DK" (not Development Kit).

ioannisg replied:
Thanks @ru-fu, that is now fixed.

galak commented:
> ...and the same for spi0, i2c0, uarte0 DTS nodes, if we want the respective examples to compile.
> 
> But _that_ triggers the duplicate address warnings you mentioned above.

We'll let's start with what default makes sense for the board.

ioannisg commented:
@galak I have changed the default board configuration so _no_ overlapping peripherals (i.e. peripherals with the same base addresses) are simultaneously enabled. I enabled:
- UARTE0, UARTE1, at ...0x8000, 0x9000
- I2C2, at 0xa000
- SPI3, at 0xb000

@anangl you might want to be notified about this.

galak commented:
@dbkinder can you review the docs.

dbkinder commented on boards/arm/nrf9160_pca10090/doc/nrf9160_pca10090.rst:
```diff
@@ -0,0 +1,207 @@
+The process requires the following steps:
+
+1. Build the Secure Zephyr application using ``-DBOARD=nrf9160_pca10090`` and
+``CONFIG_TRUSTED_EXECUTION_SECURE=y`` in the the application project configuration file.
```
This continuation line for the bullet list item needs to be indented to the ``B`` in ``Build`` on the line above.

ioannisg replied:
Fixing that, thanks.

dbkinder commented on boards/arm/nrf9160_pca10090/doc/nrf9160_pca10090.rst:
```diff
@@ -0,0 +1,207 @@
+Building a Secure only application
+==================================
+
+Build the Zephyr app in the usual way, using ``-DBOARD=nrf9160_pca10090``.
```
Do folks know what "the usual way" is?  Maybe change to:
```
Build the Zephyr app in the usual way (see :ref:`build_an_application`
and :ref:`application_run`), using ``-DBOARD=nrf9160_pca10090``.
```

ioannisg replied:
Fine, I incorporated your suggestion.

dbkinder commented on boards/arm/nrf9160_pca10090/doc/nrf9160_pca10090.rst:
```diff
@@ -0,0 +1,207 @@
+There are 2 samples that allow you to test that the buttons (switches) and LEDs on
+the board are working properly with Zephyr:
+
+.. code-block:: console
```
Replace this code-block and body with a couple of bullet item with links to the sample docs:
```
* :ref:`blinky-samle`
* :ref:`button-sample`
```

ioannisg replied:
Yes, I am fixing this as well.

ioannisg replied:
I suppose "sample" instead of "samle" :)

ioannisg commented:
Thanks for reviewing the doc @dbkinder - I addressed your comments.
